### ISSUE TITLE
Slow rendering of KaTeX on W10

### ISSUE THREAD

**Issue by ArjenB96** [laurent22/joplin#2357]
When working on my note (linked below) I notice a signifcant increase in render time when I increase the amount of KaTeX, even though I do not find it to be that much. Even after typing single letters and pausing slightly, the editor freezes for ~0.5-1 seconds to render the entire file and this disturbs my work flow.

I do not experience the issue as much on my laptop running macOS Mojave. Hiding the rendered version using the Layout button does not help the issue.

# Environment

Joplin version: 1.0.178 (prod, win32)
Platform: Windows 
OS specifics: 10, standard editor settings with 'Nord' them enabled.

## Steps To Reproduce

[slow_render.zip](https://github.com/laurent22/joplin/files/4108488/slow_render.zip)

1. Open linked file in Joplin
2. Make any (small) changes.
3. Experience significant render time.

Describe what you expected to happen:
Much less render time, allowing for a more smooth work flow.

**Comment by ArjenB96** [comment]
Log does not show anything interesting.
Furthermore, It seems that restarting the application seems to bring the render time back to acceptable levels. Not sure why?

**Comment by ArjenB96** [comment]
On start-up, Joplin uses about 175-200 MB of RAM. After working for 15-30 minutes on a single note file, it rises to 500 MB. It is at this point that the renderer becomes a bit choppy.
Even though my RAM usage is at 30% and I have 16GB of RAM available, the renderer still lags at that point.

**Comment by laurent22** [comment]
Katex code blocks are cached, so maybe the cache is malfunctioning and adding data without deleting old one.

**Comment by ArjenB96** [comment]
There are quite some instances where I use inline code blocks. Especially in the part I wrote the last hour  a significant amount of inline code blocks were written, but not a lot of text.

**Comment by ulucs** [comment]
I have been having the same issue and resorted to reinstalling the previous version as a workaround.

**Comment by laurent22** [comment]
I'm not seeing any slow down actually. Can someone post *one* example Markdown that definitely shows the problem?

**Comment by lbo34** [comment]
Hi. I am also on 1.0.178 (the app update pops for 1.0.179 but does not download it), but on OS X.
I am experiencing the same problem. The math equations are re-written every time I type something (even if it is not an equation). This happens on several documents, but I believe I did not experience this behaviour before... I am dropping an example...
[Notes de Cours Machine Learning II_ Deep Learning.md.zip](https://github.com/laurent22/joplin/files/4111636/Notes.de.Cours.Machine.Learning.II_.Deep.Learning.md.zip)

**Comment by ArjenB96** [comment]
I can confirm Ibo34's issue.
Currently trying this on OSX 10.14.5 and I have the same issue. 
Here is a screen capture of the issue:
![Joplin slow rendering of KaTeX elements](https://gfycat.com/jitterycloudygangesdolphin)

**Comment by lbo34** [comment]
I also notice some latex commands stopped working... Like \mathbb{} and some others. Used o work very well.

**Comment by lbo34** [comment]
This is bad... I tried to downgrade to 1.0.177 and I am not sure what went wrong but I cannot open the app anymore. It says it is opened but the main windows does not appear. I can access the menu but nothing happens. Only "about Joplin" seems to work and opens a window on the bottom left of the screen... I tried to reinstall 1.0.177, 1.0.178... Nothing seems to work.

**Comment by laurent22** [comment]
@lbo34 

> for instance \mathbb \reals

As this is a different issue, could you open a new issue with a complete  markdown example, and what rendering you expect?

**Comment by lbo34** [comment]
Sure... it is just these two things seems to have appeared at the same time. I'll open a new issue.

**Comment by ulucs** [comment]
Did you have any success reproducing the issue?

I also found a new workaround: disabling and then reenabling KaTeX seems to work for a while

**Comment by laurent22** [comment]
@ulucs, this workaround is to fix what issue?

**Comment by ArjenB96** [comment]
@laurent22 The lag, as it clears the RAM usage.
If I switch continuously between two medium-sized notes with KaTeX in both of them, I notice the lag showing up after a while. I have not timed it, but it might be somewhere in the range of 10-20 minutes.

**Comment by ulucs** [comment]
Yeah, it is about the lag. I also noticed the issue is also present in version 1.0.177

**Comment by laurent22** [comment]
@ArjenB96, if you could post an example of note that shows this memory leak it might help. The more sample notes we have the better.

**Comment by ArjenB96** [comment]
[lag_files.zip](https://github.com/laurent22/joplin/files/4119422/lag_files.zip)

I hope I did it right, I tried to export two notes in .jex file.
The result after 10 minutes of switching notes (while writing the summary file listed) is very significant. The resulting lag carries on when trying to edit any note.

**Comment by lbo34** [comment]
true @ulucs. it seems to help a bit.

**Comment by laurent22** [comment]
I couldn't quite replicate any slow down or memory leak on my side (it stays at 90MB on Windows and load fast), however I've optimised how the Katex assets and other plugin assets (CSS and fonts) are loaded and I think it should make it faster and consume less RAM.

Another thing that was changed in a recent release is that rendering is now asynchronous but I hope it's not the issue as that will be much harder to fix.

This change will be in the next release.
